### PR TITLE
Add enums GridRegistration and GridType for grid registration and type

### PR DESCRIPTION
**Pull request by seisman**
This PR adds the enums `GridRegistration` and `GridType` for grid registration and type.

In #3449, we added the enums `GridFormat`, but we didn't add it to the API documentation, because we didn't find a good way to display the enums (see comments in PR #3449). This PR makes the docs work by adding a new `autosummary` template for enums. 

**Preview**: 

- https://pygmt-dev--3693.org.readthedocs.build/en/3693/api/index.html#enums
- https://pygmt-dev--3693.org.readthedocs.build/en/3693/api/generated/pygmt.enums.GridRegistration.html
- https://pygmt-dev--3693.org.readthedocs.build/en/3693/api/generated/pygmt.enums.GridType.html

 
References:

- https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html
- https://rasterio.readthedocs.io/en/latest/_sources/api/rasterio.enums.rst.txt
- https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass

### CODE REVIEW

**Review by michaelgrund — APPROVED** [review]

**Comment by weiji14** [comment]
Ok with the enum API style, but before approving, I want to check that we actually want to expose the `GridFormat` enum to users? In https://github.com/GenericMappingTools/pygmt/pull/3449#issuecomment-2380598584, I thought we decided not to include `GridFormat`, but document `GridRegistration` and `GridType` (suggested in https://github.com/GenericMappingTools/pygmt/issues/499#issuecomment-2272522044)?

**Comment by seisman** [comment]
> Ok with the enum API style, but before approving, I want to check that we actually want to expose the `GridFormat` enum to users? In [#3449 (comment)](https://github.com/GenericMappingTools/pygmt/pull/3449#issuecomment-2380598584), I thought we decided not to include `GridFormat`, but document `GridRegistration` and `GridType` (suggested in [#499 (comment)](https://github.com/GenericMappingTools/pygmt/issues/499#issuecomment-2272522044))?

We don't have to document `GridFormat` here. I've cherry-picked the changes in #3696 (New enums `GridReg` and `GridType`) into this PR so that this PR can be reused.

File: pygmt/enums.py [hunk 1]
```diff
@@ -37,3 +37,21 @@ class GridFormat(IntEnum):
     GD = 22  #: Import through GDAL
     EI = 23  #: ESRI Arc/Info ASCII Grid Interchange format (ASCII integer)
     EF = 24  #: ESRI Arc/Info ASCII Grid Interchange format (ASCII float)
+
+
+class GridReg(IntEnum):
```
Comment (seisman) [inline pygmt/enums.py:42]: `GridReg` or `GridRegistration`?

Comment (weiji14) [reply pygmt/enums.py:42]: I'd prefer `GridRegistration`

Comment (seisman) [reply pygmt/enums.py]: OK, renamed to `GridRegistration` in this PR.

**Review by weiji14 — APPROVED** [review]

**Pull request merged**